src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+                                (socketAddress[26] << 16) +
+                                (socketAddress[25] << 8) +
+                                (socketAddress[24]));
+
+            return new IPAddress(address, scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        private sealed unsafe class GetAddrInfoExState : CriticalFinalizerObject, IDisposable
+        {
+            private IntPtr m_context;
+            private int m_callbackStartedOrCanceled;
+            private DnsResolveAsyncResult m_asyncResult;
+
+            public string HostAddress { get; }
+
+            public GetAddrInfoExState(string hostAddress, DnsResolveAsyncResult asyncResult)
+            {
+                HostAddress = hostAddress;
+                m_asyncResult = asyncResult;
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public void SetQueryContext(IntPtr context)
+            {
+                m_context = context;
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public bool SetCallbackStartedOrCanceled()
+            {
+                return Interlocked.Exchange(ref m_callbackStartedOrCanceled, 1) == 0;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                Task.Run(() => m_asyncResult.InvokeCallback(o));
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            ~GetAddrInfoExState()
+            {
+                // The finalizer should only be called when the AppDomain is unloading while there was a pending operation.
+                // Calling GetAddrInfoExCancel will invoke the callback inline with WSA_E_CANCELLED error code.
+
+                // If the callback is already invoked, let the traditional Dispose clear the resources.
+                // This should not be possible, because the finalizer would not have been called if the object was still alive.
+                // This is an extra protection in case the object was somehow resurrected.
+                if (!SetCallbackStartedOrCanceled())
+                    return;
+
+                ReleaseResources(cancelQuery: true);
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                ReleaseResources(cancelQuery: false);
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            private void ReleaseResources(bool cancelQuery)
+            {
+                var ptr = Interlocked.Exchange(ref m_context, IntPtr.Zero);
+
+                if (ptr == IntPtr.Zero)
+                    return;
+
+                var context = (GetAddrInfoExContext*)ptr;
+
+                if (cancelQuery)
+                    context->CancelQuery();
```

Comment:
Does the cancellation happen synchronously such that it's safe to free `context` once CancelQuery returns?

Reply:
From observation the callback is called inline with WSA_E_CANCELLED error code. But after re-reading the doc it's not clear if it's always the case.

https://msdn.microsoft.com/en-us/library/windows/desktop/hh448782(v=vs.85).aspx
>Remarks
The GetAddrInfoExCancel function cancels an asynchronous GetAddrInfoEx operation. The result is that the user's completion mechanism, either a callback or an event, is immediately invoked. No results are returned, and the error code returned for the GetAddrInfoEx asynchronous operation is set to WSA_E_CANCELLED. If the GetAddrInfoEx request has already completed or timed out, or the handle is invalid, and WSA_INVALID_HANDLE will be returned by GetAddrInfoExCancel function.

> Since many of the underlying operations (legacy name service providers, for example) are synchronous, these operations will not actually be cancelled. These operations will continue running and consuming resources. Once the last outstanding name service provider request has completed, the resources will be released.

From the wording I get the impression that the callback is always called inline and it's ok to release the context, but some internal resources may be held internally and will be released automatically when the internal sync calls actually complete. Can anyone confirms if it's really the case?

Reply:
It sounds like it should be ok, but let me ask a different question. The only time we'll cancel is if the finalizer is invoked, and the only way the finalizer would be invoked is if the AppDomain was shutting down?  In .NET Core that means the process is going away... do we need to cancel at all in that case?  We don't even run finalizers as part of shutdown any more.  Maybe the finalizer could just be used to free up resources in the rare circumstance we created one of these things and it got dropped before any operations were actually started?

Reply:
(I wrote a response to this, but it doesn't look like it took. Trying again...)

It sounds like it's ok, but let me ask a different question.  We only do this cancellation stuff if the finalizer is invoked, and the finalizer is only invoked if either a) some unexpected exception happened prior to initiating an operation or b) the AppDomain is going away, right?  There's no need to cancel for (a), and for (b), that'll only happen in core if the process is going away... plus, we don't even run finalizers on shutdown in core anymore.  Makes me wonder whether we need any of this cancellation stuff at all.

Reply:
Ohh I didn't know that finalizers were no longer called at shutdown. I think this is a problem because if the GetAddrInfoEx callback is raised after the AppDomain is unloaded, it will crash with AccessViolationException instead of a clean exit.

I just saw issue #5205, so I guess I would need to use AssemblyLoadContext.Unloading. However this complicates things because I would need to keep track of all pending operations in a static field to be able to cancel them when unloading...

Do you know any example of async p/invoke somewhere else in corefx/coreclr to see how it's handled?

Reply:
I think I was misguided here. In COREFX there will be an AppDomainUnloadedException if a native thread tries to invoke a callback from an unloaded AppDomain. But since .Net core only have the default AppDomain it might not be an issue.

I tried to trigger a crash by starting thousands of async queries before the process is shutdown, I was not able to get a crash.

You may be right and we don't need to handle cancellation at all.

Reply:
@jkotas, are you aware of any issues that might occur here if the process gets torn down while such an async operation is in flight?

I would think in .NET Core it's fine and we can just remove the cancellation stuff, and maybe add a comment indicating that more will need to be done here if this implementation is ported to netfx.

Reply:
Yes, it looks like that this is trying to deal with AppDomain unload. I think it is fine to not worry about it here. We have many of places in the framework that are incompatible with thread abort or AppDomain unload.